prometheus_metrics.go

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2024 Cloudflare, Inc. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+type PrometheusConfig struct {
+	Host       string
+	Port       string
+	ScrapePath string
+	MetricName string
+}
+
+type PrometheusMetrics struct {
+	startedAt time.Time
+	histogram prometheus.ObserverVec
+}
+
+func (p *PrometheusMetrics) Fire(result string) {
+	observer := p.histogram.With(prometheus.Labels{"method": "unknown", "status": "unknown", "result": result})
+	p.observe(observer)
+}
+
+func (p *PrometheusMetrics) ResponseStatus(method string, status int) {
+	observer := p.histogram.With(prometheus.Labels{"method": method, "status": fmt.Sprint(status), "result": "unknown"})
+	p.observe(observer)
+}
+
+func (p *PrometheusMetrics) observe(observer prometheus.Observer) {
+	elapsed := time.Now().Sub(p.startedAt)
+	observer.Observe(float64(elapsed.Milliseconds()))
+}
+
+type PrometheusMetricsFactory struct {
+	metricName string
+}
+
+func NewPrometheusMetricsFactory(config PrometheusConfig) (MetricsFactory, error) {
+	serveMux := http.NewServeMux()
+	serveMux.Handle(config.ScrapePath, promhttp.Handler())
+	server := http.Server{
+		Addr:    config.Host + ":" + config.Port,
+		Handler: serveMux,
+	}
+
+	go func() {
+		log.Printf("Listening for Prometheus scrapes on %s:%s\n", config.Host, config.Port)
+		log.Fatal(server.ListenAndServe())
```

Comment:
nits: it might be interesting to provide a context to the factory, so the server can be killed independently of the whole process.

Reply:
Interesting idea! Would this entail changes to the `MetricsFactory` interface in `metrics.go` so that there's a generic way to "stop" metrics?

Reply:
I'd say adding `Stop() error` is a way to do this, or enforcing passing a `Context`, which can then be cancelled if necessary.
It's beyond the scope of this change though.